src/nccl_ofi_rdma.c

```diff
@@ -306,65 +261,75 @@ static void unlink_topo_file()
-		 * such that it can be unlinked later. Do not write
-		 * another topology file since it would end up
-		 * overriding the stored filename. */
-		goto unlock;
-	}
-
-	if (ofi_nccl_topo_file_template()) {
-		filename = strdup(ofi_nccl_topo_file_template());
-	} else {
-		filename = strdup(topo_file_template);
-		/* Store filename to be unlinked later */
-		topo_file_unlink = filename;
+	/**
+	 * If `NCCL_TOPO_FILE` is already set, don't set it again.
+	 *
+	 * Note about forking behavior: in some Python applications, after calling
+	 * plugin init, the process will fork(). This `NCCL_TOPO_FILE` environment
+	 * variable, as well as the file descriptor it refers to, will be copied
+	 * to the child process, and will continue to point to a valid topology
+	 * file until the child process exits.
+	 */
+	if (getenv("NCCL_TOPO_FILE")) {
+		NCCL_OFI_TRACE(NCCL_INIT | NCCL_NET,
+			       "NCCL_TOPO_FILE environment variable is already set to %s",
+			       getenv("NCCL_TOPO_FILE"));
+		goto exit;
 	}
 
 	/* Create file descriptor */
-	fd = mkstemp(filename);
-	if (fd == -1) {
-		NCCL_OFI_WARN("Failed to create NCCL topology file from template %s. ERROR: %s",
-			      filename, strerror(errno));
+	topo_fd = memfd_create("ofi_nccl_topo", 0);
+	if (topo_fd == -1) {
+		NCCL_OFI_WARN("Failed to create anonymous topology file. ERROR: %s",
+			      strerror(errno));
 		ret = -errno;
-		goto unlock;
+		goto error;
 	}
 
 	/* Open file from file descriptor */
-	file = fdopen(fd, "w");
+	file = fdopen(topo_fd, "w");
 	if (file == NULL) {
-		NCCL_OFI_WARN("Failed to open NCCL topology file using file descriptor. File name: %s. ERROR %s",
-			      filename, strerror(errno));
+		NCCL_OFI_WARN("Failed to open NCCL topology file using file descriptor. ERROR %s",
+			      strerror(errno));
 		ret = -errno;
-		goto unlock;
+		goto error;
 	}
 
 	ret = nccl_ofi_topo_write(topo, file);
 	if (ret) {
-		NCCL_OFI_WARN("Failed to write NCCL topology using file descriptor. File name: %s",
-			      filename);
-		goto unlock;
+		NCCL_OFI_WARN("Failed to write NCCL topology using file descriptor. RC: %d", ret);
+		goto error;
 	}
 
-	/* Close file. The file remains accessible as long as file is not unlinked. */
-	if (fclose(file) == EOF) {
-		NCCL_OFI_WARN("Unable to close NCCL topology file. File name: %s. ERROR: %s",
-			      filename, strerror(errno));
+	/* Flush buffered writes to file. We don't close the file here so that
+	   the underlying descriptor remains open, which we will reference
+	   in `NCCL_TOPO_FILE`. */
+	if (fflush(file) == EOF) {
+		NCCL_OFI_WARN("Unable to flush NCCL topology file. ERROR: %s",
+			      strerror(errno));
 		ret = -errno;
-		goto unlock;
+		goto error;
+	}
+
+	char filename[32];
+	{
+		int size = snprintf(filename, 32, "/dev/fd/%d", topo_fd);
+		if (size < 0) {
+			NCCL_OFI_WARN("Errror preparing topo file name");
+			ret = -EIO;
+			goto error;
+		}
+		assert(size < sizeof(filename));
```

Comment:
all we're asserting here is that snprintf has well-defined behavior around its second argument, clutters the code a bit, i'd remove it.

Reply:
I hate the C standard library so much. Thanks.